addons/binding/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/net/TelnetSession.java

```diff
@@ -112,22 +126,45 @@ public void open(String host, int port) throws IOException {
 
     @Override
     public void close() throws IOException {
-        if (this.telnetClient.isConnected()) {
-            this.telnetClient.disconnect();
+        synchronized (charBuffer) {
+            logger.trace("TelnetSession close called");
+            try {
+                if (telnetClient.isConnected()) {
+                    telnetClient.disconnect();
+                }
+            } catch (IOException e) {
+                logger.debug("TelnetSession close: error disconnecting: {}", e.getMessage());
+                throw (e);
+            } finally {
+                reader = null;
```

Comment:
That's a good question.  This is using the Apache telnet client (org.apache.commons.net.telnet).  In their docs, they caution against closing the streams directly, and say that calling disconnect() is the proper way to dispose of them.  Since disconnect() is throwing an exception, we are trusting that they are doing exception handling properly and cleaning up as much as possible.  However, since the retry code will be opening a new session by calling connect() after this and retrieving the new input and output streams, I would think that any resources they may have left hanging around will be disposed of once they are no longer referenced anyway.

Reply:
I read that too but liked to ask.

Then we should follow their recommendations and do not close them. It worked before and I assume it will work afterwards and hope there will be no memory leak.

Reply:
Out of curiosity, I took a look at the source for org.apache.commons.net.telnet.  You can take a look and decide for yourself.  TelnetClient.disconnect() has exception handling that looks a bit sloppy to me, since I think the output stream will not be closed by it if there is an error closing the input stream (e.g. a buffer flush failure because of a socket error).  

http://svn.apache.org/viewvc/commons/proper/net/tags/NET_3_6/src/main/java/org/apache/commons/net/telnet/TelnetClient.java?view=markup#l113

However, it always calls super.disconnect() which resolves to SocketClient.disconnect(). This looks like it **will** actually close the input and output streams in all cases.

http://svn.apache.org/viewvc/commons/proper/net/tags/NET_3_6/src/main/java/org/apache/commons/net/SocketClient.java?view=markup#l309

As an amusing aside, SocketClient.disconnect() claims that it throws IOException, although I do not see where it would, since closeQuietly catches and does not re-throw it.

Anyway... The bottom line is that despite some strangeness in their code, I think this code is probably ok the way it is.